langchain/vectorstores/opensearch_vector_search.py

```diff
@@ -0,0 +1,366 @@
+                "script": {
+                    "source": "knn_score",
+                    "lang": "knn",
+                    "params": {
+                        "field": "vector",
+                        "query_value": query_vector,
+                        "space_type": space_type,
+                    },
+                },
+            }
+        }
+    }
+
+
+def __get_painless_scripting_source(space_type: str, query_vector: List[float]) -> str:
+    """For Painless Scripting, it returns the script source based on space type."""
+    source_value = (
+        "(1.0 + " + space_type + "(" + str(query_vector) + ", doc['vector']))"
+    )
+    if space_type == "cosineSimilarity":
+        return source_value
+    else:
+        return "1/" + source_value
+
+
+def _default_painless_scripting_query(
+    query_vector: List[float],
+    space_type: str = "l2Squared",
+    pre_filter: Dict = MATCH_ALL_QUERY,
+) -> Dict:
+    """For Painless Scripting Search, this is the default query."""
+    source = __get_painless_scripting_source(space_type, query_vector)
+    return {
+        "query": {
+            "script_score": {
+                "query": pre_filter,
+                "script": {
+                    "source": source,
+                    "params": {
+                        "field": "vector",
+                        "query_value": query_vector,
+                    },
+                },
+            }
+        }
+    }
+
+
+def _get_kwargs_value(kwargs: Any, key: str, default_value: Any) -> Any:
+    """Get the value of the key if present. Else get the default_value."""
+    if key in kwargs:
+        return kwargs.get(key)
+    return default_value
+
+
+class OpenSearchVectorSearch(VectorStore):
+    """Wrapper around OpenSearch as a vector database.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import OpenSearchVectorSearch
+            opensearch_vector_search = OpenSearchVectorSearch(
+                "http://localhost:9200",
+                "embeddings",
+                embedding_function
+            )
+
+    """
+
+    def __init__(
+        self, opensearch_url: str, index_name: str, embedding_function: Embeddings
+    ):
+        """Initialize with necessary components."""
+        self.embedding_function = embedding_function
+        self.index_name = index_name
+        self.client = _get_opensearch_client(opensearch_url)
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        refresh_index: bool = True,
+    ) -> List[str]:
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+        """
+        bulk = _import_bulk()
```

Comment:
Given this functionality shares a lot of code with from_texts, can we factor out common code into a helper method?

Reply:
Ack